graphlearn_torch/python/distributed/dist_feature.py

```diff
@@ -203,6 +236,142 @@ def _local_selecting_get(
+      remote_cnt_sum = 0
+      for ntype in ntype_list:
+        nodes = sampler_result.node.get(ntype, None)
+        if nodes is None:
+          continue
+        nodes = nodes.to(torch.long)
+        _, pb = self._get_local_store(ntype)
+        input_order= torch.arange(nodes.size(0),
+                                  dtype=torch.long,
+                                  device=self.device)
+        partition_ids = pb[nodes.to(pb.device)].to(self.device)
+        nodes = nodes.to(self.device)
+        if pidx == self.partition_idx:
+          continue
+        else:
+          remote_mask = (partition_ids == pidx)
+          remote_ids = torch.masked_select(nodes, remote_mask)
+          indexes[ntype][pidx] = torch.masked_select(input_order, remote_mask)
+          ssize = remote_ids.numel()
+          send_ids[offset: offset + ssize] = remote_ids.tolist()
+          remote_cnt_sum = remote_cnt_sum + remote_ids.numel()
+          offset = offset + ssize
+      remote_cnt_list[pidx] = remote_cnt_sum
+
+    assert len(send_ids) == sum(remote_cnt_list)
+    send_sr, recv_sr = communicate_node_num(remote_cnt_list)
+    _, trecv = sum(send_sr), sum(recv_sr)
+    self.recv_rn_count = []
+    for pidx in range(self.num_partitions):
+      self.recv_rn_count.append(int(recv_sr[pidx]))
+    self.recv_rn_gnid = torch.zeros(trecv, dtype=torch.long)
+    dist.all_to_all_single(self.recv_rn_gnid, torch.tensor(send_ids),
+                           self.recv_rn_count, remote_cnt_list.tolist(),
+                           async_op=False)
+    return remote_cnt_list, indexes
+
+  def communicate_node_feats(
+    self,
+    ntype_list: List[NodeType],
+    remote_cnt: torch.Tensor,
+    send_num_dict: Dict[NodeType, int],
+    recv_num_dict: Dict[NodeType, int],
+    indexes: Dict[NodeType, List]
+  ):
+    rfeats_list = []
+    offset = 0
+    for ntype in ntype_list:
+      feat_num = recv_num_dict.get(ntype)
+      if feat_num > 0:
+        feat, _ = self._get_local_store(ntype)
+        ntype_ids = self.recv_rn_gnid[offset:offset+feat_num]
+        offset = offset + feat_num
+        rfeats_list.append(feat[ntype_ids])
+    rfeats_send = torch.cat(rfeats_list, dim=0)
+    feat_size = rfeats_send.shape[1]
+    send_count = self.recv_rn_count
+    recv_count = remote_cnt.tolist()
+    recv_feats = torch.zeros((sum(recv_count), feat_size), dtype=rfeats_send.dtype)
+    req = dist.all_to_all_single(recv_feats, rfeats_send,
+                                 recv_count, send_count,
+                                 async_op=True)
+    req.wait()
+    dist.barrier()
+    recv_feat_list = torch.split(recv_feats, recv_count, dim = 0)
+    remote_feats_dict = {}
+    for ntype in ntype_list:
+      remote_feats_dict[ntype] = []
+
+    for pidx in range(self.num_partitions):
+      if pidx == self.partition_idx:
+        continue
+      else:
+        offset = 0
+        for ntype in ntype_list:
+          send_num = send_num_dict.get(ntype)
```

Comment:
Sorry I pasted my last question at a wrong line number...
My question here is: why different partitions are using the same `send_num_dict`? My understanding is that `send_num_dict` stores the total number of node IDs of a specific node type that the host partition requests from the others.

Reply:
Yes, `send_num_dict` stores the total number of node IDs of a specific node type that the host partition requests from the others. However, for each rank/partition, it is an independent process, the `send_num_dict` variable is defined in this process, and it is not a global variable. For example, the `send_num_dict` variable in partition 0 is different from  the `send_num_dict` variable in rank 1. I print this variable and get this:
In Partition 0:
`pidx: 0, send_num_dict: {'author': tensor(12330), 'fos': tensor(6682), 'institute': tensor(354), 'paper': tensor(13632)}`
and In partition 1:
`pidx: 1, send_num_dict: {'author': tensor(12022), 'fos': tensor(6377), 'institute': tensor(328), 'paper': tensor(13317)}`

Reply:
Within the for loop between ln 349 and ln 359, we are iterating all the partitions and apply this line `ntype_feat = recv_feat_list[pidx][offset:offset+send_num, :]` to fetch the received feature data. It seems for a specific node type, we are using the same `send_num = send_num_dict.get(ntype)` for all the partitions. But as our above discussions, `send_num_dict` stores the total number of node IDs of a specific node type that the host partition requests from the others, not just for a single partition.

Reply:
Within the for loop between ln 349 and ln 359, we are iterating all the partitions and apply this line `ntype_feat = recv_feat_list[pidx][offset:offset+send_num, :]` to fetch the received feature data. It seems for a specific node type, we are using the same `send_num = send_num_dict.get(ntype)` for all the partitions. But as our above discussions, `send_num_dict` stores the total number of node IDs of a specific node type that the host partition requests from the others, not just for a single partition.

Reply:
Within the for loop between ln 349 and ln 359, we are iterating all the partitions and apply this line `ntype_feat = recv_feat_list[pidx][offset:offset+send_num, :]` to fetch the received feature data. It seems for a specific node type, we are using the same `send_num = send_num_dict.get(ntype)` for all the partitions. But as our above discussions, `send_num_dict` stores the total number of node IDs of a specific node type that the host partition requests from the others, not just for a single partition.

Reply:
Oops, here should be a mistake. We need to maintain a `Dict[NodeType, List[int]] ` here for variable `send_num_dict` and `recv_num_dict`. Sorry I just validated this PR using 2 partitions for igbh-small dataset, so didn't find the bug...

Reply:
Thanks, could you please resolve the conflicts of this PR?